### PR TITLE
Add playwright e2e tests

### PR DESCRIPTION
**Pull request by owi92**
This adds the playwright framework for e2e UI tests as outlined in #855.

Several tests are already included but are still subject to change and improve. Unfortunately some of these can be flaky and are prone to getting stuck when loading and or navigating (in like 10% of all runs). I'm unsure if that is caused by poor test design or outside conditions. Right now the config allows for 2 retries to be extra sure everything passes, but in in my experience 1 retry should suffice. 
**Edit**: Actually, I haven't encountered any test failures in CI so I changed that back to 1 retry.

Especially the `realm` and `block` tests might still benefit from breaking them down into smaller chunks, but since so much of these tests depend on one another, I started by doing them in one swoop and in sequence.

I broke each test down into several steps and sub-steps using `test.step`, but there are definitely other ways of doing this. The benefit of using `test.step` is mainly for readability and sequentiality (is that even a word? google says it is 🤷)

Another thing to consider for these tests:
Right now they rely on certain events and series to be present _somewhere_ in Tobira, and at least one series block to be present on the "welcome page". I believe this is suboptimal since that data might change in the future (correct me if I'm wrong of course).

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (LukasKalbertodt) [inline .github/workflows/ci.yml:139]: Does it make sense to always upload the results? Feels like it only makes sense if the test failed? Or what is this good for?

File: .github/workflows/ci.yml
Comment (LukasKalbertodt) [inline .github/workflows/ci.yml:144]: Oh! I didn't know you can specify retention days here! That could be a solution to our other CI problem. Will think about it again!

File: .github/workflows/ci.yml
Comment (LukasKalbertodt) [inline .github/workflows/ci.yml:133]: I think the `echo yes |` is not needed here. You probably had `rebuild` here earlier, which does require that. But `update` does not.

File: frontend/tests/languageSelection.spec.ts
Comment (LukasKalbertodt) [inline frontend/tests/languageSelection.spec.ts:39]: Missing trailing newline. Our CI should catch that :/

File: frontend/tests/languageSelection.spec.ts
Comment (LukasKalbertodt) [inline frontend/tests/languageSelection.spec.ts:7]: Wouldn't it be better to locate these things via something that is also something a user would use? E.g. `title` attribute must contain "Language selection"?

File: frontend/tests/search.spec.ts
Comment (LukasKalbertodt) [inline frontend/tests/search.spec.ts:27]: Missing trailing newline

File: frontend/tests/videoPage.spec.ts
Comment (LukasKalbertodt) [inline frontend/tests/videoPage.spec.ts:87]: missing trailing newline

File: frontend/tests/search.spec.ts
Comment (LukasKalbertodt) [inline frontend/tests/search.spec.ts:8]: Same as a previous comment: I think it would be better to just select "an input where the placeholder is 'search'" or sth like that. At least my very little knowledge about UI tests tells me to code them as close to what a user would do.

File: frontend/tests/search.spec.ts
Comment (LukasKalbertodt) [inline frontend/tests/search.spec.ts:12]: Nothing wrong with this, just reminded me: we certainly need tests of typing something in a different input that contains `s` and making sure in that case, it does NOT jump to the search bar.

**Review by LukasKalbertodt — CHANGES_REQUESTED** [review]
Finally some tests :tada: 

This is just a first review: I have not yet looked at all the files in `tests`, but at everything else. Most of these comments are mainly questions for me to understand decisions :D After resolving all of that I can have another look at the actual tests.

File: .github/workflows/ci.yml
Comment (LukasKalbertodt) [reply .github/workflows/ci.yml:144]: So yeah I think we should do the following: `ci.yaml` should mark its artifact (not this playwright one, but the main Tobira one) with `retention-days: 1`. Then `deploy.yml` can download it as usually. Well actually, we can simplify that script as we don't need to use custom JS code anymore, but can just use a standard action.
And then, playwright testing should be moved to a new CI job that runs after this `main` job and also downloads the artifact. 
With only 1 day of retention, we shouldn't run into any limit problems.

But I'd suggest doing all of this as a new PR after this one is merged.

File: frontend/playwright.config.ts
Comment (owi92) [reply frontend/playwright.config.ts:9]: > Why would one want to do that?

You mean restricting wokers on CI? According to https://playwright.dev/docs/ci#workers, it's to "prioritize stability and reproducibility".

> And why do we always restrict us to one worker?

The tests for `blocks` and `realms` create and rely on specific realms. When any of these run at the same time in chromium and firefox, that can lead to race conditions.
Unfortunately, it's not possible to configure worker usage for a single test file yet, as far as I can tell.

Hmm but I suppose I might be able to circumvent this by using different realms based on which browser the test is run in - that should be possible. Will look into that.

File: frontend/playwright.config.ts
Comment (owi92) [reply frontend/playwright.config.ts:19]: Yeah, these were created automatically. Do we even want to keep them? If so, I can of course change the commenting style.

File: frontend/playwright.config.ts
Comment (owi92) [reply frontend/playwright.config.ts:64]: I think it did time out on me once or twice, that's why I increased the limit. But maybe I'm misremembering. I'll try it without the limit.

File: frontend/tests/languageSelection.spec.ts
Comment (owi92) [reply frontend/tests/languageSelection.spec.ts:7]: Hm I did that at first but then thought that these selectors maybe shouldn't rely on the translation file. When the translation gets changed at some point, the selector also needs to be changed accordingly or otherwise the test will fail.

On the other hand the `_react` selector doesn't work or rather can't be used on everything, so you'll see me using text content as locator in other places anyway. Meaning we'll need to check the tests anyway when changing translations. 
I'll leave it as is for now, but I'm very open to replacing the `_react` selectors again. They definitely aren't something that makes writing these tests easier (though in some cases they might still be useful, e.g. when an element does not contain any text... whiiich is seldom the case so that point is also kind of moot).

File: frontend/playwright.config.ts
Comment (LukasKalbertodt) [reply frontend/playwright.config.ts:9]: Thanks for the explanation! 

And yeah, the race conditions seem kind of unfortunate. I mean... ideally every test would get its own fresh database and be completely isolated. But I'm sure setting this up is toooons of work. So if you find a way to avoid race conditions in a simply way as you suggest, why not go for it. That said, don't invest too much time into this. Running locally with 1 worker is certainly fine for now.

File: frontend/playwright.config.ts
Comment (LukasKalbertodt) [reply frontend/playwright.config.ts:19]: Uhm. I'll let you decide what of those comments are worth keeping and which are not. And for those we keep, I would change the comment style, yes.

File: .github/workflows/ci.yml
Comment (LukasKalbertodt) [reply .github/workflows/ci.yml:139]: Ah no, your explanation makes sense. And it's a really tiny file anyway, so let's keep it like this. But maaaybe add a very brief comment there to explain it? Not super important tho.

**Comment by owi92** [comment]
Dangit I pushed the rebase combined with my changes again 🤦.
Though I think I would suggest reviewing this file-by-file anyway.

File: frontend/playwright.config.ts
Comment (LukasKalbertodt) [reply frontend/playwright.config.ts]: So you changed it locally to `undefined`. So the race condition stuff is solved?

File: frontend/tests/languageSelection.spec.ts
Comment (owi92) [reply frontend/tests/languageSelection.spec.ts]: I replaced most of the `_react` selectors but some are still required to find elements without any text, such as series blocks. We could also give these a `data-testid` prop but that basicallly does the same as the react selector. Though the react selector is marked as experimental, so it might be useful to swap these anyway.
(Using a test id looks sth like `<SeriesBlock data-testid="seriesBlock">` and then inside our tests use `page.getByTestId("seriesBlock")`, if want you can also take a look at the [docs](https://playwright.dev/docs/locators#locate-by-test-id).)

**Comment by narickmann** [comment]
I took a look at the tests.

The tests are looking good so far and I think they cover pretty much everything. 
I haven't noticed anything important that's still missing.

**Review by LukasKalbertodt — APPROVED** [review]
Then let's just merge this! I still haven't done a full review due to time contraints, but that's fine. This is not code that is shipped anyway. The worst that could happen is a bit of flaky tests in CI, with random CI failures. But that's easy to fix and no big problem. No need to block this any longer. Makes more sense to test-drive these tests on `master` anyway.

**Pull request merged**
